rpc/src/client/transport/websocket.rs

```diff
@@ -650,38 +652,69 @@ impl WebSocketClientDriver {
             return Ok(());
         }
 
-        let wrapper = match serde_json::from_str::<response::Wrapper<GenericJsonResponse>>(&msg) {
+        let wrapper: response::Wrapper<GenericJsonResponse> = match serde_json::from_str(&msg) {
             Ok(w) => w,
             Err(e) => {
                 error!(
                     "Failed to deserialize incoming message as a JSON-RPC message: {}",
                     e
                 );
+
                 debug!("JSON-RPC message: {}", msg);
+
                 return Ok(());
             }
         };
+
+        debug!("Generic JSON-RPC message: {:?}", wrapper);
+
         let id = wrapper.id().to_string();
+
+        if let Some(e) = wrapper.into_error() {
+            self.publish_error(&id, e).await;
+        }
+
         if let Some(pending_cmd) = self.pending_commands.remove(&id) {
-            return self.respond_to_pending_command(pending_cmd, msg).await;
+            self.respond_to_pending_command(pending_cmd, msg).await?;
         };
+
         // We ignore incoming messages whose ID we don't recognize (could be
         // relating to a fire-and-forget unsubscribe request - see the
         // publish_event() method below).
         Ok(())
     }
 
+    async fn publish_error(&mut self, id: &str, err: Error) {
```

Comment:
There's a bit of duplicated code between `publish_error` and `publish_event`, we should probably clean this up